### PR TITLE
Added Weam and Crosswalk

### PR DESCRIPTION
**Pull request by mphprogrammer**
Added Crosswalk and some additional Weam CSV stuff.

### CODE REVIEW

**Comment by ayaleloehr** [comment]
@mphprogrammer you looking for a review on this, or is this a WIP?

**Comment by mphprogrammer** [comment]
It's a "reviewable wip" that I am moving specs in for
 Perhaps it would be good to get some eyes on now

**Comment by mphprogrammer** [comment]
However, there are pieces in place for later work which may not make much sense right now, but do take a peek.

File: db/migrate/20161209204400_create_versions.rb [hunk 1]
```diff
@@ -1,12 +1,14 @@
 class CreateVersions < ActiveRecord::Migration
   def change
     create_table :versions do |t|
-      t.belongs_to :user, index: true, null: false
-      
-      t.integer :version, index: true, null: false
-      t.boolean :production, index: true, default: false, null: false
```
Comment (ayaleloehr) [inline db/migrate/20161209204400_create_versions.rb]: @mphprogrammer production used to have an index but doesn't anymore. is this intentional?

File: .rubocop.yml [hunk 2]
```diff
@@ -24,9 +24,14 @@ Metrics/LineLength:
...
 
 Metrics/MethodLength:
-  Enabled: true
   Max: 20
 
+Metrics/MethodLength:
+  Max: 120
+
+Metrics/ModuleLength:
+  Max: 120
```
Comment (saneshark) [inline .rubocop.yml]: I think the defaults for these are actually good. It encourages better abstraction, breaking things out into separate modules or classes, and making sure that methods and cyclomatic complexity of methods are not too extreme, making code more readable.

File: Gemfile [hunk 3]
```diff
@@ -47,6 +47,7 @@ gem 'figaro'
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug'
+  gem 'pry'
```
Comment (saneshark) [inline Gemfile]: this should be included automatically as part of pry-rails?

File: app/assets/javascripts/crosswalks.js [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
+// All this logic will automatically be available in application.js.
```
Comment (saneshark) [inline app/assets/javascripts/crosswalks.js]: not sure why these stubs are necessary.

File: app/assets/stylesheets/crosswalks.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Crosswalks controller here.
+// They will automatically be included in application.css.
```
Comment (saneshark) [inline app/assets/stylesheets/crosswalks.scss]: not sure why these stubs are necessary.

File: app/controllers/crosswalks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+class CrosswalksController < ApplicationController
+end
```
Comment (saneshark) [inline app/controllers/crosswalks_controller.rb]: not sure why these stubs are necessary.

File: app/assets/javascripts/crosswalks.js [hunk 7]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (ayaleloehr) [inline app/assets/javascripts/crosswalks.js]: @mphprogrammer I want to verify that you really want a crosswalks.js and crosswalks.scss file?

File: app/helpers/crosswalks_helper.rb [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+module CrosswalksHelper
+end
```
Comment (saneshark) [inline app/helpers/crosswalks_helper.rb]: not sure why these stubs are necessary.

File: app/models/concerns/standardizable.rb [hunk 9]
```diff
@@ -54,22 +54,26 @@ def override_setters
...
     end
 
     def override_facility_code_setter
+      klass = self
+
       define_method 'facility_code=' do |value|
-        value = value.try(:strip)
-        self['facility_code'] = value ? value.upcase.rjust(8, '0') : nil
+        value = value.try(:strip).try(:upcase)
+        self['facility_code'] = klass.forbidden_word?(value) || value.blank? ? nil : value.rjust(8, '0')
```
Comment (saneshark) [inline app/models/concerns/standardizable.rb]: This doesn't seem right at all, you're using a method to dynamically define a new method? I'm not sure why this is necessary, but it's definitely a code smell.

File: app/models/concerns/standardizable.rb [hunk 10]
```diff
@@ -54,22 +54,26 @@ def override_setters
...
       end
     end
 
     def override_institution_setter
       define_method 'institution=' do |value|
-        value = value.try(:strip)
-        self['institution'] = value ? value.upcase : nil
+        value = value.try(:strip).try(:upcase)
+        self['institution'] = value ? value : nil
```
Comment (saneshark) [inline app/models/concerns/standardizable.rb]: here as well, dynamically defining a method using meta programming doesn't seem necessary, perhaps I'm missing context, but this is just confusing.

File: app/models/concerns/standardizable.rb [hunk 11]
```diff
@@ -80,6 +84,28 @@ def override_state_setter
       end
     end
 
+    def override_ope_setter
+      klass = self
+
+      define_method('ope=') do |value|
```
Comment (saneshark) [inline app/models/concerns/standardizable.rb]: and here.

File: app/models/concerns/standardizable.rb [hunk 12]
```diff
@@ -80,6 +84,28 @@ def override_state_setter
...
+
+      define_method('ope6') do
+        ope.present? ? ope[1, 5] : nil
+      end
+    end
+
+    def override_cross_setter
+      klass = self
+
+      define_method('cross=') do |value|
```
Comment (saneshark) [inline app/models/concerns/standardizable.rb]: and here

File: app/models/concerns/standardizable.rb [hunk 13]
```diff
@@ -89,6 +115,7 @@ def override_generic_setter(col, data_type)
 
     def override_generic_boolean_setter(col)
       klass = self
+
       define_method "#{col}=" do |value|
```
Comment (saneshark) [inline app/models/concerns/standardizable.rb]: and here

**Review by saneshark — COMMENTED** [review]
So this is just a first pass, it looks like many of the files are just CSVs so naturally I didn't review those.

I will probably want to take a look at the migrations and specs more closely, but I think based on the feedback I'm giving here some things will likely change.

Particularly the use of metaprogramming, `define_method` seems like the wrong approach. I might be missing something related to this, but from what I'm seeing here it doesn't seem necessary.

File: app/assets/javascripts/crosswalks.js (hunk 7 above)
Comment (mphprogrammer) [inline app/assets/javascripts/crosswalks.js]: @ayaleloehr I was going to remove all the *.js, *.scss , helper files, etc in one fell swoop... But I can do that now...

File: app/assets/javascripts/crosswalks.js (hunk 4 above)
Comment (mphprogrammer) [inline app/assets/javascripts/crosswalks.js]: Going away in another PR, as there will be 23 of them...

File: app/assets/stylesheets/crosswalks.scss (hunk 5 above)
Comment (mphprogrammer) [inline app/assets/stylesheets/crosswalks.scss]: Going away in another PR, as there will be 23 of them...

File: app/controllers/crosswalks_controller.rb (hunk 6 above)
Comment (mphprogrammer) [inline app/controllers/crosswalks_controller.rb]: They will be incorporated in a later PR ...

File: app/helpers/crosswalks_helper.rb (hunk 8 above)
Comment (mphprogrammer) [inline app/helpers/crosswalks_helper.rb]: Helpers going away in a later PR ...

**Comment by mphprogrammer** [comment]
Terminating branch.

**Pull request closed without merging**